No chop and no split deletions

cshannon opened the pull request:
This commit changes deletions to no longer require splitting a tablet or chop compactions. Instead, tablets that overlap the start/end of a deletion range will be have its files fenced to exclude the rows that are part of the deletion.

Instead of splitting tablets for deletion, the metadata gets updated to create either 1 or 2 ranges in the tablet to fence files. 1 range is created if the deletion covers part of the tablet at the start or end. 2 ranges are created if the deletion is in the middle of the tablet to fence off the middle. If splits are not needed but the the tablet needs to be chopped (there is a boundary shared between a tablet at start/end an a tablet being deleted) then a fence is also created to avoid chop compactions.

This is a follow on to #3640  and this closes #3679

cshannon committed:
No chop and no split deletions

This commit changes deletions to no longer require splitting a tablet or
chop compactions. Instead, tablets that overlap the start/end of a
deletion range will be have its files fenced to exclude the rows that
are part of the deletion.

cshannon commented:
I started a full IT build and will report back the results when finished.

cshannon commented:
> I started a full IT build and will report back the results when finished.

The full IT build passed.

ctubbsii commented on server/manager/src/main/java/org/apache/accumulo/manager/TabletGroupWatcher.java:
```diff
@@ -924,6 +1015,135 @@ private void mergeMetadataRecords(MergeInfo info) throws AccumuloException {
     }
   }
 
+  private boolean needsFencingForDeletion(MergeInfo info, KeyExtent keyExtent) {
+    return needsFencingForSplit(info, keyExtent) || info.deleteOverlaps(keyExtent);
+  }
```
This is the only method that calls needsFencingForSplit. needsFencingForSplit also double-checks that the operation `isDelete()`, but that's redundant if this method is being called.

Putting the contents of that method into this one, and simplifying it dramatically, I get:

```suggestion
  private boolean needsFencingForDeletion(MergeInfo info, KeyExtent keyExtent) {
    // Does this extent cover the end points of the delete?
    Predicate<Text> isWithin = r -> r != null && keyExtent.contains(r);
    Predicate<Text> isNotBoundary =
        r -> !r.equals(keyExtent.endRow()) && !r.equals(keyExtent.prevEndRow());
    KeyExtent deleteRange = info.getExtent();
    return (keyExtent.overlaps(deleteRange) && Stream
        .of(deleteRange.prevEndRow(), deleteRange.endRow()).anyMatch(isWithin.and(isNotBoundary)))
        || info.deleteOverlaps(keyExtent);
  }
```

Now, I'm still not exactly sure about the edge cases this is checking for... there are too many overlaps, contains, and boundary checks happening in different methods to really intuitively get it. But this simplification should help. I selected predicate names that suggest what is being checked here, so that it would be easier to understand why you might need fencing in some cases, but not in others. You don't need fencing, for example, if the delete range's endRow is within the current keyExtent, but falls exactly on its boundary... you can just delete all the files. But you would need to fence if the delete range is within the current keyExtent, but doesn't reach all the way at the boundary.

I'm really not sure about the correctness of this when it comes to the edge cases, though, like the first or last tablet (where endRow or prevEndRow are null), or if the bounds of the deleted range are not inclusive. I'm hoping there's already good coverage of all this, so you don't have to do much... but simplification like what I am suggesting here could make it much more readable if done in more places.

cshannon replied:
Thanks I will take a look when I get a chance and try and see if it's correct and the changes make sense. It's tricky as the logic combines checks for both when we need to do chop compactions and also if we need to split in order to see if we need to fence the files instead. Also once things are working and we have good test coverage we can always then clean up or improve the code and business logic. I figure some of this will be improved a bit over time as we test and find ways to optimize.

cshannon replied:
My latest commit uses this suggestion, it seems to be equivalent after looking at it and is a lot more concise. All the tests still pass.

ctubbsii commented on server/manager/src/main/java/org/apache/accumulo/manager/TabletGroupWatcher.java:
```diff
@@ -557,6 +560,35 @@ private int assignedOrHosted(Map<TableId,TableCounts> last) {
     return result;
   }
 
+  // This method is used to detect if a tablet needs to be split for a delete
+  // Instead of performing a split, the tablet will have it's files fenced.
+  private boolean needsFencingForSplit(MergeInfo info, KeyExtent extent) {
+    // Merges don't split
+    if (!info.isDelete()) {
+      return false;
+    }
+    // Does this extent cover the end points of the delete?
+    KeyExtent range = info.getExtent();
+    if (extent.overlaps(range)) {
+      for (Text splitPoint : new Text[] {range.prevEndRow(), range.endRow()}) {
+        if (splitPoint == null) {
+          continue;
+        }
+        if (!extent.contains(splitPoint)) {
+          continue;
+        }
+        if (splitPoint.equals(extent.endRow())) {
+          continue;
+        }
+        if (splitPoint.equals(extent.prevEndRow())) {
+          continue;
+        }
+        return true;
+      }
+    }
+    return false;
+  }
+
```
This method appears to only be used in one place, and can be consolidated into that method. I do that in my other comment, so this method can just be deleted.

```suggestion
```

ctubbsii commented on server/manager/src/main/java/org/apache/accumulo/manager/state/MergeStats.java:
```diff
@@ -136,17 +128,10 @@ public MergeState nextMergeState(AccumuloClient accumuloClient, CurrentState man
-        } else {
-          state = MergeState.WAITING_FOR_CHOPPED;
-        }
+        // Todo: Clean up references to WAITING_FOR_CHOPPED and SPLITTING and remove
+        // from enum in a future PR as both are going away. for now just change
```
See my other comment about serialization of the ordinals, and the risks of removing these enums.

ctubbsii reviewed (changes requested):
This code is a nightmare to look at and work with. Not your code specifically... just the totality of what's there, due to the complexity of all the edge cases and the overlaps checks, and the deep call stacks that are difficult to follow; I'm not even sure what can be done to clean any of it up. Your code thankfully is not making it worse... but I do wonder if there's more we can do to make some of these internals more maintainable in general.

cshannon commented on server/manager/src/main/java/org/apache/accumulo/manager/Manager.java:
```diff
@@ -649,16 +648,6 @@ TabletGoalState getGoalState(TabletLocationState tls, MergeInfo mergeInfo) {
             case STARTED:
             case SPLITTING:
-              return TabletGoalState.HOSTED;
-            case WAITING_FOR_CHOPPED:
-              if (tls.getState(tserverSet.getCurrentServers()).equals(TabletState.HOSTED)) {
-                if (tls.chopped) {
-                  return TabletGoalState.UNASSIGNED;
-                }
-              } else if (tls.chopped && tls.walogs.isEmpty()) {
-                return TabletGoalState.UNASSIGNED;
-              }
-
```
One of the things I wasn't sure about was if we were just going to leave all the chop compaction related stuff around as deprecated even if not used anymore or just delete it all. I figured removing may cause issues with upgrades or other problems but I wasn't sure since chop compaction code should mostly be internal in the servers and I wouldn't expect to touch public API a lot (but I haven't checked), although there is a Thrift api for running a chop compaction. In the elasticity branch I figure we will delete it all but in main I guess we can keep it and mark as deprecated.

I was going to tackle the deprecation/removal in a follow on PR (upgrading metadata would also be a follow on PR) so I can revert back the things I removed such as this switch statement for now. We also need to handle the upgrade case in a follow on PR to upgrade existing metadata to the new format.

ctubbsii replied:
Definitely delete what isn't needed, if it's internal code. The only thing you need to be careful about keeping around is stuff related to serializing (if it's persisted anywhere needed across upgrades), or if it's public API. Then, we need to be more careful.

keith-turner replied:
> Unfortunately, we (inconveniently) chose to use the ordinal to serialize the merge info for fate operations, so we have to be very careful not to cause a problem by changing the ordinals and interpreting them incorrectly across an upgrade. If we hadn't done that (which, in general, is a bad idea), then we could have just removed it entirely, since it's only used internally, and we expect all the jars we use to be upgraded at the same time.

Accumulo used to not upgrade if there are outstanding FATE ops. Reasoning about FATEs operations running across an upgrade is not easy to do, its also really difficult to test.  So we should be able to remove this enum.  If Accumulo is no longer preventing upgrade when there are outstanding FATE ops, then I would consider that a bug.

keith-turner replied:
Found the code that does the check for outstanding fate ops in the upgrade code.

https://github.com/apache/accumulo/blob/770c05b46a4e3c135c85d0772ebb2b02d9db5414/server/manager/src/main/java/org/apache/accumulo/manager/upgrade/UpgradeCoordinator.java#L157

cshannon commented:
> This code is a nightmare to look at and work with. Not your code specifically... just the totality of what's there, due to the complexity of all the edge cases and the overlaps checks, and the deep call stacks that are difficult to follow; I'm not even sure what can be done to clean any of it up. Your code thankfully is not making it worse... but I do wonder if there's more we can do to make some of these internals more maintainable in general.

@ctubbsii - Thanks for the review, you are right this was quite a challenge to work through all the edge cases and I'm not sure if I got them all or not. I got the ones that of course were failing due to testing but there may be more that might come up. When I was working through the issues the main test case I used to help was https://github.com/apache/accumulo/blob/5bec1c33af6c464ce75f25ade7b4bd4dba0709ea/test/src/main/java/org/apache/accumulo/test/functional/DeleteRowsIT.java#L82-L115 

That test seemed to cover most stuff but I need to think about if there are more. Something else that I just realized I still need to test but didn't have time for yet is deleting ranges when tablets already contained fenced files which makes sure clipping of ranges works. (I do have tests for this already for the merge case though). A simple way to test that would be to merge some tablets to create fenced files (and make sure compaction is disabled so it doesn't clean it up after) and then run a delete that will touch the files again so we have to narrow the ranges on a file that already has a range. I can do that later this week to make sure it works and if it doesn't fix any issues.

One nice thing about putting all of this into the no-chop-merge branch is it gives an opportunity to keep testing and do follow up PRs before doing one big final review. I think things are finally getting close though so hoping we can get this merged into main before too long.

keith-turner commented:
While looking at this locally I noticed that `TabletGroupWatcher.sendSplitRequest()` is no longer used.  Could that be removed?  Also could `MergeState.SPLITTING` be removed?

keith-turner commented on server/manager/src/main/java/org/apache/accumulo/manager/TabletGroupWatcher.java:
```diff
@@ -924,6 +1015,135 @@ private void mergeMetadataRecords(MergeInfo info) throws AccumuloException {
+
+      // Find the tablets that overlap the start and end of the deletion range
+      final List<TabletMetadata> tabletMetadatas = new ArrayList<>();
+      loadTabletMetadata(range.tableId(), startRow, ColumnType.PREV_ROW, ColumnType.FILES)
+          .ifPresent(tabletMetadatas::add);
+      loadTabletMetadata(range.tableId(), lastRow, ColumnType.PREV_ROW, ColumnType.FILES)
+          .ifPresent(tabletMetadatas::add);
```
If the delete range falls within in a single tablet, would the same tablet end up tabletMetadatas twice?  If so would that cause any problems in the loop below?

cshannon replied:
If the delete range is in the same tablet you end up with two metadata files, we have to fence twice. So you end up deleting the previous file metadata and replacing it with two new files. The first file would have a range that covers the start of the tablet to the start of the delete range and the second file would have a range that covers the end of the delete range to the end of the tablet so that the entire delete range in the middle of the tablet is fenced off and not visible.

keith-turner replied:
Will the code end up deleting the original files twice in the metadata table?  That should be ok I think, tyring to make sure I understand the code. Would it be ok to deduplicate tablets before the loop and if there is the same tablet twice only go through the loop once?

cshannon replied:
Hmm I wonder if I can use a set or a map to track the TabletMetadata, that way if we have the same thing twice it won't run twice.

cshannon replied:
The latest commit fixes this to make sure we don't process a tablet twice.

cshannon commented on server/manager/src/main/java/org/apache/accumulo/manager/TabletGroupWatcher.java:
```diff
@@ -669,8 +671,97 @@ private void updateMergeState(Map<TableId,MergeStats> mergeStatsCache) {
+      // If needsFencingForDeletion() is true then this tablet was previously fenced
+      // so we need to keep it and not delete the tablet. Skip to the next highest
+      // tablet after this one if it exists
+      if (firstTablet != null && needsFencingForDeletion(info, firstTablet)) {
```
You are right, this check is not needed, it was just force of habit of checking for null and I will get rid of it. I can also get rid of it for looking up the last tablet, something should always return since it's using the overlapping() method for looking up tablets so rows should always map to something.

cshannon commented:
> While looking at this locally I noticed that `TabletGroupWatcher.sendSplitRequest()` is no longer used. Could that be removed? Also could `MergeState.SPLITTING` be removed?

We can definitely remove sendSplitRequest() as it's private and I just missed that as an oversight. For the Enum change, based on your other comment that outstanding FATE operations should prevent upgrade, it sounds like we can remove that as well as it should not be a problem.

cshannon committed:
Clean up code from feedback

cshannon committed:
Merge branch 'no-chop-merge' into no-chop-no-split-delete

keith-turner commented on server/manager/src/main/java/org/apache/accumulo/manager/TabletGroupWatcher.java:
```diff
@@ -924,6 +1015,135 @@ private void mergeMetadataRecords(MergeInfo info) throws AccumuloException {
+            // should be exclusive for the start row.
+            for (Range fenced : ranges) {
+              // Clip range with the tablet range if the range already exists
+              fenced = existing.hasRange() ? existing.getRange().clip(fenced) : fenced;
```
The ranges could be disjoint which would throw an exception,  thinking if they are disjoint then just drop the file w/o adding anything?  It would be cleaner to put the rest of the code in an else instead of doing a continue like the code below, was just trying to make the suggestion short.

```suggestion
              fenced = existing.hasRange() ? existing.getRange().clip(fenced, true) : fenced;

              if(fenced == null) {
                m.putDelete(DataFileColumnFamily.NAME, existing.getMetadataText());
                continue;
              }
```

cshannon replied:
I actually just ran into this exact situation as I was testing deleting with tablets that already have ranged files so I will be making sure to include this with my next testing updates.

cshannon replied:
This has been fixed and includes new tests in https://github.com/apache/accumulo/pull/3728/commits/0e6c5d70bc1fad6b234206e103d1851d52e752df

If a file is disjoint for all the new range(s) then it will be dropped. There maybe 2 ranges created if the deletion range is contained within 1 tablet so we need to check both. the ITs include tests for deleting existing ranged files

keith-turner commented on server/manager/src/main/java/org/apache/accumulo/manager/TabletGroupWatcher.java:
```diff
@@ -924,6 +1015,135 @@ private void mergeMetadataRecords(MergeInfo info) throws AccumuloException {
+  // to exclude the portion of the tablet that should be deleted
+  private List<Range> createRangesForDeletion(TabletMetadata tabletMetadata,
+      final KeyExtent deleteRange) {
+    final KeyExtent keyExtent = tabletMetadata.getExtent();
```
Not completely sure the following assumption is true, but think it might be.  Still looking through the code. If the assumption is true it would be good to add the validation.

```suggestion
   // its assumed that the delete range overlaps this range but does not contain it.
    Preconditions.checkArgument(!deleteRange.contains(tabletMetadata.getExtent()));

    final KeyExtent keyExtent = tabletMetadata.getExtent();
```

keith-turner commented on server/manager/src/main/java/org/apache/accumulo/manager/TabletGroupWatcher.java:
```diff
@@ -924,6 +1015,135 @@ private void mergeMetadataRecords(MergeInfo info) throws AccumuloException {
+
+    // This covers the case of when a deletion range overlaps the first tablet
+    // We need to create a range that excludes the deletion
+    if ((firstKey == null || !firstKey.equals(deleteRange.toDataRange().getStartKey()))
+        && deleteRange.toDataRange().getStartKey() != null
+        && tabletRange.contains(deleteRange.toDataRange().getStartKey())) {
+      Manager.log.trace(
+          "Fencing tablet at start of deletion range with end row {}, Start deletion row {}",
+          prevTabletEndRow, deleteRange.toDataRange().getStartKey().getRow());
+      ranges.add(new Range(prevTabletEndRow, false,
+          deleteRange.toDataRange().getStartKey().getRow(), false));
+    }
```
Assuming that for the case where the first row of the delete range is null that it means everything in the tablet would be deleted from the perspective of the prev row.  If that is true then would not want to act when firstKey is null and could simplify the code to the following.  Not completely about this, also I did not adjust the log message.

```suggestion
    if (firstKey != null && tabletMetadata.getExtent().contains(firstKey.getRow())) {
      // In the case where firstKey is null, there is no data to keep in the tablet.
      Manager.log.trace(
          "Fencing tablet at start of deletion range with end row {}, Start deletion row {}",
          prevTabletEndRow, deleteRange.toDataRange().getStartKey().getRow());
      ranges.add(new Range(prevTabletEndRow, false, deleteRange.prevEndRow(), true));
    }
```

cshannon replied:
I didn't test this but I'm assuming the changes here likely won't work for the same reasons as commented here: https://github.com/apache/accumulo/pull/3728#discussion_r1313531334

keith-turner commented on server/manager/src/main/java/org/apache/accumulo/manager/TabletGroupWatcher.java:
```diff
@@ -924,6 +1015,135 @@ private void mergeMetadataRecords(MergeInfo info) throws AccumuloException {
+
+    // This covers the case of when a deletion range overlaps the last tablet
+    // We need to create a range that excludes the deletion.
+    if ((tabletEndRow == null
+        || !tabletEndRow.equals(deleteRange.toDataRange().getEndKey().getRow()))
+        && deleteRange.toDataRange().getEndKey() != null
+        && tabletRange.contains(deleteRange.toDataRange().getEndKey())) {
+      Manager.log.trace(
+          "Fencing tablet at end of deletion range with end row {}, End deletion row {}",
+          tabletEndRow, deleteRange.toDataRange().getEndKey().getRow());
+      ranges.add(new Range(deleteRange.toDataRange().getEndKey().getRow(), tabletEndRow));
+    }
```
The following does nothing for a null end row and as a result can simplify the code.  Also did not adjust the log message.

```suggestion
    if (deleteRange.endRow() != null && tabletMetadata.getExtent().contains(deleteRange.endRow())) {
      // in the case where the delete range end row is null, there not data to keep in the tablet.
      Manager.log.trace(
          "Fencing tablet at end of deletion range with end row {}, End deletion row {}",
          tabletEndRow, deleteRange.toDataRange().getEndKey().getRow());
      ranges.add(new Range(deleteRange.endRow(), false, tabletEndRow, true));
    }
```

cshannon replied:
I don't think the change here works. All the statements I added in the if clause were necessary (or at least _seemed_ to be but I could be wrong) and done as I found different edge cases when running the DeleteRowsIT test. if you make this change the tests don't work. 

I would need to go and see exactly which of the test cases fail for all the conditions I added as I worked through them one at a time. Specifically for the case of a null endRow, we need to include the range if you are only partially deleting the last tablet. You are right that if the entire last tablet is being deleted then it wouldn't need to be part of a range but if we are only deleting part of the last tablet we need to include it in the range list to be fenced. If you get rid of tabletEndRow == null in the condition then when you run the test it hangs and breaks on this line: https://github.com/apache/accumulo/blob/3409d958bb1d69baf5b93bb8a588a3e3ee9aab1c/test/src/main/java/org/apache/accumulo/test/functional/DeleteRowsIT.java#L111

When I get a chance I can take a look and dig into it more for the different conditions buut the simplest way to verify is to make changes to the logic and then run the tests in `DeleteRowsIT`.

keith-turner replied:
ok, I will try running the test and experimenting with the change

keith-turner commented on server/manager/src/main/java/org/apache/accumulo/manager/TabletGroupWatcher.java:
```diff
@@ -924,6 +1015,135 @@ private void mergeMetadataRecords(MergeInfo info) throws AccumuloException {
+
+      // Find the tablets that overlap the start and end of the deletion range
+      final List<TabletMetadata> tabletMetadatas = new ArrayList<>();
+      loadTabletMetadata(range.tableId(), startRow, ColumnType.PREV_ROW, ColumnType.FILES)
```
For this what is the case where the tablet is not present?

cshannon replied:
I don't think there is a case where it wouldn't find a tablet. the loadTabletMetadata() method calls off to Ample and uses the overlapp() method so it should always map and return a matching tablet.

keith-turner reviewed (commented):
Not finished reviewing this, posted the comments I have so far.   Some of the comments were made with a partial understanding of these change so could be incorrect.  Plan to circle back later and review more of the changes.

keith-turner commented on server/manager/src/main/java/org/apache/accumulo/manager/TabletGroupWatcher.java:
```diff
@@ -924,6 +967,135 @@ private void mergeMetadataRecords(MergeInfo info) throws AccumuloException {
+
+          // Create the ranges for fencing the files, this takes the place of
+          // chop compactions and splits
+          final List<Range> ranges = createRangesForDeletion(tabletMetadata, range);
```
Can this return an empty list and if so what should happen?

cshannon replied:
Hmm, I don't think it would ever be empty because we are only iterating over the 1 or 2 tablets that overlap the deletion range by looking up the tablets previously. I would need to think about it but at least in the existing tests it's never empty and should always be at least 1 range or 2 (if the deletion is contained in 1 tablet). We may want to add a check for this to throw an error as I don't think it would be valid if it was empty.

keith-turner replied:
One case I was unsure about was what happens when deleting null,null, basically all data.   Was not sure if this code gets involved in that case.

cshannon replied:
That case works and is handled and is covered by DeleteRowsIT. In this case it never hits that block of code because the [line](https://github.com/cshannon/accumulo/blob/64484c420ec5c61e3d7dcfbb47c4b4bcfe88ab16/server/manager/src/main/java/org/apache/accumulo/manager/TabletGroupWatcher.java#L1025) that checks if the tablet needs fencing will be false.

When it runs with null, null the returned tablet that overlaps is `1:a<;` and that fails the needs fencing check so it never processes anything and [logs](https://github.com/cshannon/accumulo/blob/64484c420ec5c61e3d7dcfbb47c4b4bcfe88ab16/server/manager/src/main/java/org/apache/accumulo/manager/TabletGroupWatcher.java#L1084) a statement. No tablets are fenced and later all tablets get deleted. This check ends up being false and you see the following in the log when running a test:

`2023-09-01T17:18:38,785 [manager.Manager] DEBUG: Skipping metadata update on file for keyExtent 1;a< for delete as not overlapping on rows`

keith-turner replied:
Thanks for the info it will save me time when I get back to reviewing this, I had not looked at `needsFencingForDeletion` yet.

cshannon committed:
Handle disjoint ranges for existing ranged files on deletion

cshannon committed:
Address more formatting comments

cshannon commented on server/base/src/main/java/org/apache/accumulo/server/manager/state/MergeInfo.java:
```diff
@@ -81,25 +81,22 @@ public boolean isDelete() {
+      return false;
+    }
+  }
+
```
Fixed in latest commit

cshannon committed:
Simplify the needsFencingForDeletion method

Combines existing methods and makes things simpler for checking if a
tablet needs to have its files fenced during a deletion

---------

Co-authored-by: Christopher Tubbs <ctubbsii@apache.org>

cshannon committed:
Optimize to make sure we don't need to process a tablet twice during
deletes

Also simplify the handling of disjoint ranges and file deletions

cshannon committed:
formatting and added check for empty range list

cshannon committed:
add back inadvertent removal of null check

cshannon commented:
@keith-turner - Ok I think you can take a look at this again, I pushed a bunch of updates today and tried to address most of the comments so far.

keith-turner commented on server/manager/src/main/java/org/apache/accumulo/manager/TabletGroupWatcher.java:
```diff
@@ -924,6 +942,168 @@ private void mergeMetadataRecords(MergeInfo info) throws AccumuloException {
+
+  // Instead of splitting or chopping tablets for a delete we instead create ranges
+  // to exclude the portion of the tablet that should be deleted
+  private List<Range> createRangesForDeletion(TabletMetadata tabletMetadata,
+      final KeyExtent deleteRange) {
+    final KeyExtent keyExtent = tabletMetadata.getExtent();
+    final Text prevTabletEndRow =
+        tabletMetadata.getPrevEndRow() != null ? tabletMetadata.getPrevEndRow() : null;
+    final Text tabletEndRow = keyExtent.endRow();
+    final Range tabletRange = tabletMetadata.getExtent().toDataRange();
+    final Key firstKey =
+        prevTabletEndRow != null ? new Key(prevTabletEndRow).followingKey(PartialKey.ROW) : null;
+    final List<Range> ranges = new ArrayList<>();
+
+    // This covers the case of when a deletion range overlaps the first tablet
+    // We need to create a range that excludes the deletion
+    if ((firstKey == null || !firstKey.equals(deleteRange.toDataRange().getStartKey()))
+        && deleteRange.toDataRange().getStartKey() != null
+        && tabletRange.contains(deleteRange.toDataRange().getStartKey())) {
+      Manager.log.trace(
+          "Fencing tablet at start of deletion range with end row {}, Start deletion row {}",
+          prevTabletEndRow, deleteRange.toDataRange().getStartKey().getRow());
+      ranges.add(new Range(prevTabletEndRow, false,
+          deleteRange.toDataRange().getStartKey().getRow(), false));
+    }
+
+    // This covers the case of when a deletion range overlaps the last tablet
+    // We need to create a range that excludes the deletion.
+    if ((tabletEndRow == null
+        || !tabletEndRow.equals(deleteRange.toDataRange().getEndKey().getRow()))
+        && deleteRange.toDataRange().getEndKey() != null
+        && tabletRange.contains(deleteRange.toDataRange().getEndKey())) {
+      Manager.log.trace(
+          "Fencing tablet at end of deletion range with end row {}, End deletion row {}",
+          tabletEndRow, deleteRange.toDataRange().getEndKey().getRow());
+      ranges.add(new Range(deleteRange.toDataRange().getEndKey().getRow(), tabletEndRow));
+    }
+
+    return ranges;
+  }
```
I was experimenting with the following locally and was able to get DeleteRowsIT to pass. I find this easier to understand.

```suggestion
  private Text followingRow(Text row) {
    if (row == null)
      return null;
    return new Key(row).followingKey(PartialKey.ROW).getRow();
  }

  // Instead of splitting or chopping tablets for a delete we instead create ranges
  // to exclude the portion of the tablet that should be deleted
  private List<Range> createRangesForDeletion(TabletMetadata tabletMetadata,
      final KeyExtent deleteRange) {
    final KeyExtent tabletExtent = tabletMetadata.getExtent();

    // If the delete range wholly contains the tablet being deleted then there is no range to clip
    // files to because the files should be completely dropped.
    Preconditions.checkArgument(!deleteRange.contains(tabletExtent), "delete range:%s tablet:%s",
        deleteRange, tabletExtent);

    final List<Range> ranges = new ArrayList<>();

    if (deleteRange.overlaps(tabletExtent)) {
      if (deleteRange.prevEndRow() != null
          && tabletExtent.contains(followingRow(deleteRange.prevEndRow()))) {
        Manager.log.trace("Fencing tablet {} files to ({},{}]", tabletExtent,
            tabletExtent.prevEndRow(), deleteRange.prevEndRow());
        ranges.add(new Range(tabletExtent.prevEndRow(), false, deleteRange.prevEndRow(), true));
      }

      // This covers the case of when a deletion range overlaps the last tablet. We need to create a
      // range that excludes the deletion.
      if (deleteRange.endRow() != null
          && tabletMetadata.getExtent().contains(deleteRange.endRow())) {
        Manager.log.trace("Fencing tablet {} files to ({},{}]", tabletExtent, deleteRange.endRow(),
            tabletExtent.endRow());
        ranges.add(new Range(deleteRange.endRow(), false, tabletExtent.endRow(), true));
      }
    } else {
      Manager.log.trace(
          "Fencing tablet {} files to itself because it does not overlap delete range",
          tabletExtent);
      ranges.add(tabletExtent.toDataRange());
    }

    return ranges;
  }
```

cshannon replied:
The changes here look good to me so I went ahead and merged it, it is a lot simpler by adding the overlaps() check

keith-turner commented on server/manager/src/main/java/org/apache/accumulo/manager/TabletGroupWatcher.java:
```diff
@@ -669,8 +598,97 @@ private void updateMergeState(Map<TableId,MergeStats> mergeStatsCache) {
+
   private void deleteTablets(MergeInfo info) throws AccumuloException {
-    KeyExtent extent = info.getExtent();
+    // Before deleting actual tablets we need to fence the files in tablets
```
Still have not completed reviewing this.  Looking through what I have so far I am wondering if it would simplify things to merge the new and old code into a single scan of the metadata table that does everything.  This single scan could keep state as it goes and avoid having to recompute things. Was thinknig of something like the following.

```java
  private void deleteTablets(MergeInfo info) throws Exception {
    Preconditions.checkArgument(info.isDelete());
    var deleteRange = info.getExtent();

    // TODO hard coded table name
    try (var tablets = manager.getContext().getAmple().readTablets().forTable(deleteRange.tableId())
        .overlapping(deleteRange.prevEndRow(), false, null).checkConsistency().saveKeyValues()
        .build(); var writer = manager.getContext().createBatchWriter("accumulo.metadata")) {
      KeyExtent firstTabletDeleted = null;
      for (var tablet : tablets) {
        var tabletExtent = tablet.getExtent();
        if (deleteRange.contains(tabletExtent)) {
          // delete tablet because its completely contained in delete range
          deleteTablet(writer, tablet);

          if (firstTabletDeleted == null) {
            firstTabletDeleted = tabletExtent;
          }
        } else if (deleteRange.overlaps(tabletExtent)) {
          // this could be the first tablet, last tablet, or both first and last tablet
          fenceTablet(writer, deleteRange, tablet);

          if (deleteRange.endRow() != null && tabletExtent.contains(deleteRange.endRow())) {
            // this is the last tablet
            if (firstTabletDeleted != null) {
              // TODO could fence set prevend row in same mutation that fences
              // some tablets were deleted so need to reset this tablets prev end row
              setPrevEndRow(writer, tablet, firstTabletDeleted.prevEndRow());
            }
            // saw the last tablet and fixed it up, so done
            break;
          }
        } else {
          // this tablet is disjoint with the deletion range AND we did not see a last tablet that
          // partially overlaps, so need to fence it and set its prev end row
          fenceTablet(writer, deleteRange, tablet);

          // adjust the tablets prev end row
          setPrevEndRow(writer, tablet, firstTabletDeleted.prevEndRow());
        }
      }
    }
  }
```

While writing this I realized this code needs to be idempotent and the above code is not idempotent, but I think could be made so with some slight modifications.  Do you know if the changes you have made so far are idempotent?

cshannon replied:
I can take a look at implementing something like this where we change the code to just do one pass of the metadata vs the current version which does 2 passes (first updates/fences and then the deletion of the tablets). Instead of deleting tablets another option is to just drop the files and run the merge command after as a follow on. I can explore these various things in a separate branch and follow on since they are big changes.

I am not sure if the current version is idempotent, I would need to look at it and think about it in terms of what happens in failure scenarios.

cshannon committed:
Simply createRangesForDeletion method

Co-authored-by: Keith Turner <kturner@apache.org>

cshannon committed:
Merge branch 'no-chop-merge' into no-chop-no-split-delete

cshannon committed:
Add missing braces to fix errorpone warning

cshannon committed:
Merge branch 'no-chop-merge' into no-chop-no-split-delete

keith-turner commented on server/manager/src/main/java/org/apache/accumulo/manager/TabletGroupWatcher.java:
```diff
@@ -669,8 +598,97 @@ private void updateMergeState(Map<TableId,MergeStats> mergeStatsCache) {
+    return deletionEndRow;
+  }
+
+  private boolean hasTabletsToDelete(MergeInfo info, Text deletionStartRow, Text deletionEndRow) {
+    // If the endRow of the deletion range is not null, but the actual deletionEndRow is null then
+    // this means
+    // the first tablet was fenced so there are no tablets to delete and we can return
+    if (info.getExtent().endRow() != null && deletionEndRow == null) {
+      return false;
+      // If the prevEndRow of the deletion range is not null, but the actual deletionStartRow is
+      // null then this means
+      // the last tablet was fenced so there are no tablets to delete and we can return
+    } else if (info.getExtent().prevEndRow() != null && deletionStartRow == null) {
+      return false;
+      // Make sure detected deletionEndRow is after deletionStartRow. If they are the same tablet or
+      // before then nothing to delete and we can return
+      // This would happen in some situations such as a deletion range fully contained in 1 tablet
+    } else if (deletionEndRow != null && deletionStartRow != null
+        && new Key(deletionEndRow).compareRow(deletionStartRow) <= 0) {
+      return false;
+    }
+
+    return true;
```
Are the following the only two case where there are not tablet to delete?

 1. The case where delete range falls completely within a single tablet.
 2. The case where the delete range overlaps two contiguous tablets.  For example a table with splits `[c,m,s]` and a delete range of `(f,q]` then the tablet with end row `m` and `s` would each be fenced.  So those two tablets are contiguous with nothing between them.

For both cases above is the assumption that deletionStartRow and deletionEndRow are equal?  I think that is the case.


I am wondering if this code could do the following instead where that code handles nulls. But not sure about this, just thinking out loud.  

```suggestion
  private boolean hasTabletsToDelete(MergeInfo info, Text deletionStartRow, Text deletionEndRow) {
     return !Objects.equals(deletionStartRow, deletionEndRow);
  }
```

Also I do not understand the case in last else if it where it does ``new Key(deletionEndRow).compareRow(deletionStartRow) <= 0`, not sure about this less than or equals check and how it fits into the bigger picture.

cshannon replied:
@keith-turner - I don't think your simplification entirely works (I tried it and it fails with errors), It's been a couple weeks but I know I added the checks here for the various edge cases I encountered and with comments about how, however it's possible that this could be re-worked so it is easier to understand and maybe a slight tweak to what you did will work.

I started working through this a bit again using the comments I left as a guide and I am going back through the tests to look at the different edge cases that caused these scenarios and I can report back what deletion range caused each one and why.  After that we can look at if there is a better way.

cshannon replied:
@keith-turner - Ok I went through this method and I updated it in https://github.com/apache/accumulo/pull/3728/commits/f1887245213c65f1a415c60e0a39818fb53e37a3 to be a lot more clear and simpler. 

There are 4 conditions where we don't have any tablets to delete and only fencing is needed:

1. If the first and last tablets that overlap the deletion range are equal then the range is inside a single tablet that is fenced.
2. If the last tablet that overlaps the deletion range is the first tablet in the table then there is nothing before it to delete and that tablet is fenced.
3. If the first tablet that overlaps the deletion range is the last tablet in the table then there is nothing after it to delete and that tablet is fenced.
4. If the first and last tablet that overlap the deletion range are contiguous then both are fenced and nothing in between to delete.

cshannon replied:
I pushed one more commit to rename things to be more clear: https://github.com/apache/accumulo/pull/3728/commits/0d744b9b2fa1a05a4904606629c1da78d64f80b0

keith-turner replied:
Those changes are really nice.

keith-turner replied:
@cshannon  curious if the test cover all the cases you mentioned above.

cshannon replied:
> @cshannon curious if the test cover all the cases you mentioned above.

Yep, the DeleteRowsIT covers all the cases, I went through it when I made my latest changes to make sure everything looked good. There's a test that eliminates some rows in only one tablet, one that eliminates some rows in the first tablet, another one that eliminates some rows in the last tablet and a test that causes two tablets to be contiguous and fenced with nothing in the middle. I can double check though to see if we need more coverage. I'm planning follow on PRs anyways for more test coverage (i really want to exercise and make sure all edge cases are good)

cshannon commented on server/manager/src/main/java/org/apache/accumulo/manager/TabletGroupWatcher.java:
```diff
@@ -924,6 +942,176 @@ private void mergeMetadataRecords(MergeInfo info) throws AccumuloException {
+            }
+
+            // If the existingFile is not contained in the newFiles set then we can delete it
+            Sets.difference(existingFile, newFiles).forEach(
```
Unless I'm not understanding the use case you had in mind I think the behavior of dropping the file in the disjoint case should be correct.  First the valid range(s) for the files are computed after processing the delete range (ie the range we are not deleting and should be visible still). Each existing file is fenced with those new range(s) and get updated. The exception being if a file already has an existing range when then need to clip its existing range with the new range(s) to see if there's any overlap. If disjoint then the file does not have any rows that are part of the new range(s) after deletion so the file should be dropped entirely.

keith-turner replied:
> Unless I'm not understanding the use case you had in mind I think the behavior of dropping the file in the disjoint case should be correct. 

oh yeah that is correct.  I was getting the ranges mixed up, I was thinking of the overall deletion range.  This range is not the deletion range but the range outside the deletion range inside of a tablet.

cshannon commented on server/manager/src/main/java/org/apache/accumulo/manager/TabletGroupWatcher.java:
```diff
@@ -924,6 +942,176 @@ private void mergeMetadataRecords(MergeInfo info) throws AccumuloException {
+              "No ranges found that overlap deletion range.");
+
+          // Go through and fence each of the files that are part of the tablet
+          for (Entry<StoredTabletFile,DataFileValue> entry : tabletMetadata.getFilesMap()
```
So there are 2 parts to this, first whether it can happen and second if it is an issue.

In terms of can it happen, off the top of my head I'm not 100% sure but I would think this may be possible depending on previous merge or deletions with things being clipped. I think to avoid it we'd need to make sure to collapse ranges on merging/deletes when we do metadata updates. Right now the merge code updates files and copies existing files from merged tablets independently but doesn't try and collapse ranges or anything for duplicate files. It was easier to do the merging without trying to collapse ranges for now but could look at doing that. The fencing RFile iterator does collapse a list of ranges though.

In terms of if it's an issue, I don't think this scenario would cause any actual problems. If you end up with two files that are identical it would just do an update/add with an identical file and range but it would just replace so wouldn't end up with a duplicate or anything. Is that correct? If this is an issue then one option would be to move the tracking of the two sets of files (the newFiles and existingFile sets) earlier and create them before the for loop that iterates over the files for the tablet and just track all new/deleted files across entire tablet before doing mutations instead of resetting the sets for each file. My reason for doing it for each file was to keep memory usage lower and not have to track files for the entire tablet all at once.

keith-turner replied:
> If you end up with two files that are identical it would just do an update/add with an identical file and range but it would just replace so wouldn't end up with a duplicate or anything. Is that correct? 

That does seem correct, I can not think of any serious problems caused by this,  There could be information loss in the data file value, which is not a serious problem.

Thinking about this a bit more, I can not think of situation with merge where it would produce files with overlapping ranges.  When merge adds ranges to files, its based on tablet ranges which are disjoint.   When this new code adds 2 ranges to the same file in a tablet, those 2 ranges are disjoint.  So maybe its not something that would be expected to happen, still not sure though.

cshannon replied:
Yeah, I can try and think if there's a scenario but if it does happen we don't lose the data at least and ranges are the same.

cshannon commented on server/manager/src/main/java/org/apache/accumulo/manager/TabletGroupWatcher.java:
```diff
@@ -669,8 +598,97 @@ private void updateMergeState(Map<TableId,MergeStats> mergeStatsCache) {
+    if (deletionStartRow != null) {
+      Manager.log.trace("Finding tablet that contains the start row {} for deletion",
+          deletionStartRow);
+      final KeyExtent firstTablet =
+          loadTabletMetadata(info.getExtent().tableId(), deletionStartRow, ColumnType.PREV_ROW)
+              .map(TabletMetadata::getExtent).orElse(null);
+      // If needsFencingForDeletion() is true then this tablet was previously fenced
+      // so we need to keep it and not delete the tablet. Skip to the next highest
+      // tablet after this one if it exists
+      if (needsFencingForDeletion(info, firstTablet)) {
+        Key nextExtent = new Key(deletionStartRow).followingKey(PartialKey.ROW);
+        deletionStartRow = getHighTablet(new KeyExtent(info.getExtent().tableId(),
+            nextExtent.getRow(), info.getExtent().prevEndRow())).endRow();
```
I don't think this is quite right but I am working on updating the `getDeletionEndRow` and `deletionStartRow` methods to simply things and borrowing part of this

cshannon replied:
@keith-turner  - I have simplified things here with the getDeletionStartRow() and getDeletionEndRow() methods in my latest commit https://github.com/apache/accumulo/pull/3728/commits/a778403e20c5fc2148f71b2e693122bd5d47f453

cshannon committed:
Simplify getDeletionStartRow and getDeletionEndRow

cshannon commented on server/manager/src/main/java/org/apache/accumulo/manager/TabletGroupWatcher.java:
```diff
@@ -924,6 +942,176 @@ private void mergeMetadataRecords(MergeInfo info) throws AccumuloException {
+      final SortedMap<KeyExtent,TabletMetadata> tabletMetadatas = new TreeMap<>();
+      loadTabletMetadata(range.tableId(), startRow, ColumnType.PREV_ROW, ColumnType.FILES)
+          .ifPresent(tm -> tabletMetadatas.put(tm.getExtent(), tm));
+      loadTabletMetadata(range.tableId(), lastRow, ColumnType.PREV_ROW, ColumnType.FILES)
```
In https://github.com/apache/accumulo/pull/3728/commits/a778403e20c5fc2148f71b2e693122bd5d47f453 I have updated this to no longer load the tablet metadata if the row is null as we don't need to do fencing in that case.

cshannon committed:
Simplify the hasTabletsToDelete method

cshannon committed:
Rename methods and variables for better clarity

keith-turner commented on server/manager/src/main/java/org/apache/accumulo/manager/TabletGroupWatcher.java:
```diff
@@ -1004,28 +1027,43 @@ private List<Range> createRangesForDeletion(TabletMetadata tabletMetadata,
+      // to fence a starting tablet as we are deleting everything up to the end tablet
+      // Likewise, if the endRow is null there will be an empty endTablet as we are deleting
+      // all tablets after the starting tablet
+      final Optional<TabletMetadata> startTablet =
+          Optional.ofNullable(startRow).flatMap(er -> loadTabletMetadata(range.tableId(), startRow,
+              ColumnType.PREV_ROW, ColumnType.FILES));
+      final Optional<TabletMetadata> endTablet = Optional.ofNullable(endRow).flatMap(
+          er -> loadTabletMetadata(range.tableId(), endRow, ColumnType.PREV_ROW, ColumnType.FILES));
```
If the following is correct, would be a bit cleaner to use the lambda parameter.

```suggestion
      final Optional<TabletMetadata> startTablet =
          Optional.ofNullable(startRow).flatMap(row -> loadTabletMetadata(range.tableId(), row,
              ColumnType.PREV_ROW, ColumnType.FILES));
      final Optional<TabletMetadata> endTablet = Optional.ofNullable(endRow).flatMap(
          row -> loadTabletMetadata(range.tableId(), row, ColumnType.PREV_ROW, ColumnType.FILES));
```

cshannon replied:
Thanks for catching that, that was an oversight on my part.

cshannon committed:
minor code cleanup and test updates

cshannon committed:
Merge branch 'no-chop-merge' into no-chop-no-split-delete
